vaadinfrontend/src/main/java/life/qbic/datamanager/views/navigation/ProjectSideNavigationComponent.java

```diff
@@ -0,0 +1,206 @@
+import life.qbic.projectmanagement.domain.project.experiment.ExperimentId;
+import org.slf4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * Project Side Navigation Component
+ * <p>
+ * Allows the user to switch between the components shown in each {@link ProjectMainLayout} by
+ * clicking on the corresponding {@link SideNavItem} within Component which routes the user to the
+ * respective route defined in {@link life.qbic.datamanager.views.AppRoutes} for the component in
+ * question.
+ */
+
+@SpringComponent
+@UIScope
+public class ProjectSideNavigationComponent extends Div implements
+    BeforeEnterObserver {
+
+  private static final Logger log = getLogger(ProjectSideNavigationComponent.class);
+  private final Div projectSection = new Div();
+  private final SideNavItem projectSectionHeader = new SideNavItem("");
+  private final Div experimentSection = new Div();
+  private final SideNavItem experimentSectionHeader = new SideNavItem("");
+  private final transient ProjectInformationService projectInformationService;
+  private final transient ExperimentInformationService experimentInformationService;
+  public static final String PROJECT_ID_ROUTE_PARAMETER = "projectId";
+  public static final String EXPERIMENT_ID_ROUTE_PARAMETER = "experimentId";
+  private transient Context context = new Context();
+  private final MenuBar projectSelectMenu = new MenuBar();
+
+  public ProjectSideNavigationComponent(
+      @Autowired ProjectInformationService projectInformationService,
+      @Autowired ExperimentInformationService experimentInformationService) {
+    Objects.requireNonNull(projectInformationService);
+    Objects.requireNonNull(experimentInformationService);
+    this.addClassName("project-navigation-drawer");
+    this.projectInformationService = projectInformationService;
+    this.experimentInformationService = experimentInformationService;
+    projectSection.addClassName("project-section");
+    experimentSection.addClassName("experiment-section");
+    add(projectSection, experimentSection);
+    log.debug(
+        "New instance for ProjectSideNavigationComponent {} was created",
+        System.identityHashCode(this));
+  }
+
+  @Override
+  public void beforeEnter(BeforeEnterEvent beforeEnterEvent) {
+    resetSections();
+    String projectId = beforeEnterEvent.getRouteParameters()
+        .get(PROJECT_ID_ROUTE_PARAMETER).orElseThrow();
+    ProjectId parsedProjectId = ProjectId.parse(projectId);
+    this.context = new Context().with(parsedProjectId);
+    beforeEnterEvent.getRouteParameters().get(EXPERIMENT_ID_ROUTE_PARAMETER)
+        .ifPresent(experimentId -> this.context = context.with(ExperimentId.parse(experimentId)));
+    setContext();
+  }
+
+  private void setContext() {
+    initializeProjectSection(context.projectId().orElseThrow());
+    initializeExperimentSection(context.projectId().orElseThrow());
+  }
+
+  private void initializeProjectSection(ProjectId projectId) {
```

Comment:
this method introduces side effects (manipulating project section header and project section). Please retrieve the required project information once, and pass required information to the methods that will build and return a configured side nav element. 
On top level, you can then compose the nav bar, with the advantage that your code is not as nested (which makes reading and debugging hard) and you have fewer side effects. Try to remove any side effect in your functions, as they easily lead to unexpected behaviour. 
Also, your function does more than one thing. Can you spot it? ;) Please honor the definition of a function.

Reply:
I rewrote how the section are initialized to remove their interdependency as much as possible and load the experiment/project information beforehand 👍